File: Patterns/Abstract Factory Pattern/Cpp/Pizza/PepperoniPizza.cpp
#pragma once

#include "./Pizza.cpp"
#include "./Ingredient/PizzaIngredientFactory.cpp"

class PepperoniPizza : public Pizza {
public:
    PizzaIngredientFactory* ingredientFactory;

    PepperoniPizza(PizzaIngredientFactory* ingredientFactory) : ingredientFactory{ingredientFactory} {}

    void prepare() override {
        cout << "준비 중: " << name;

        dough = ingredientFactory->createDough();
        sauce = ingredientFactory->createSauce();
        pepperoni = ingredientFactory->createPepperoni();

        cout << dough->getName() << "를 돌리는 중...\n";
        cout << sauce->getName() << "를 뿌리는 중...\n";
        cout << "토핑을 올리는 중...\n";
        cout << '\t' << pepperoni->getName() << '\n';
    }
};

File: Patterns/Abstract Factory Pattern/CSharp/Pizza/PepperoniPizza.cs
using System;

using PizzaIngredientFactory = Pizza.Ingredient.PizzaIngredientFactory;

namespace Pizza
{
    public class PepperoniPizza : Pizza
    {
        private PizzaIngredientFactory ingredientFactory;

        public PepperoniPizza(PizzaIngredientFactory ingredientFactory)
        {
            this.ingredientFactory = ingredientFactory;
        }

        public override void Prepare()
        {
            Console.WriteLine($"준비 중: {Name}");

            dough = ingredientFactory.CreateDough();
            sauce = ingredientFactory.CreateSauce();
            pepperoni = ingredientFactory.CreatePepperoni();

            Console.WriteLine($"{dough?.Name}을 뿌리는 중...");
            Console.WriteLine($"{sauce?.Name}을 뿌리는 중...");
            Console.WriteLine("토핑을 올리는 중...");
            Console.WriteLine($"\t{pepperoni?.Name}");
        }
    }
}